internal/provider/providers/scaleway/provider.go

```diff
@@ -0,0 +1,191 @@
+	return p.ipVersion
+}
+
+func (p *Provider) IPv6Suffix() netip.Prefix {
+	return p.ipv6Suffix
+}
+
+func (p *Provider) Proxied() bool {
+	return false
+}
+
+func (p *Provider) BuildDomainName() string {
+	return utils.BuildDomainName(p.owner, p.domain)
+}
+
+func (p *Provider) HTML() models.HTMLRow {
+	return models.HTMLRow{
+		Domain: fmt.Sprintf("<a href=\"http://%s\">%s</a>", p.BuildDomainName(), p.BuildDomainName()),
+		Owner:  p.Owner(),
+		Provider:  "<a href=\"https://www.scaleway.com/\">Scaleway</a>",
+		IPVersion: p.ipVersion.String(),
+	}
+}
+
+// Update updates the DNS record for the domain using Scaleway's API.
+// API documentation: https://www.scaleway.com/en/developers/api/domains-and-dns/#path-records-update-records-within-a-dns-zone
+func (p *Provider) Update(ctx context.Context, client *http.Client, ip netip.Addr) (newIP netip.Addr, err error) {
+    // Construct the URL for the API request
+    u := url.URL{
+        Scheme: "https",
+        Host:   "api.scaleway.com",
+        Path:   fmt.Sprintf("/domain/v2beta1/dns-zones/%s/records", p.domain),
+        RawQuery: fmt.Sprintf("A=%s", ip.String()),
+    }
+
+    // Prepare the request body
+    requestBody := map[string]interface{}{
+        "changes": []map[string]interface{}{
+            {
+                "set": map[string]interface{}{
+                    "id_fields": map[string]interface{}{
+                        "name": p.field_name,
+                        "type": p.field_type,
+                    },
+                    "records": []map[string]interface{}{
+                        {
+                            "data": ip.String(),
+                            "ttl":  p.ttl,
+                        },
+                    },
+                },
+            },
+        },
+    }
+    requestBodyBytes, err := json.Marshal(requestBody)
+    if err != nil {
+        return netip.Addr{}, fmt.Errorf("json marshal: %w", err)
+    }
+
+    // Create the HTTP request
+    request, err := http.NewRequestWithContext(ctx, http.MethodPatch, u.String(), bytes.NewReader(requestBodyBytes))
+    if err != nil {
+        return netip.Addr{}, fmt.Errorf("creating http request: %w", err)
+    }
+    request.Header.Set("Content-Type", "application/json")
+    request.Header.Set("Accept", "application/json")
+    request.Header.Set("X-Auth-Token", p.secretKey)
+    headers.SetUserAgent(request)
+
+    // Send the request
+    response, err := client.Do(request)
+    if err != nil {
+        return netip.Addr{}, fmt.Errorf("doing http request: %w", err)
+    }
+    defer response.Body.Close()
+
+    // Read and clean the response body
+	s, err := utils.ReadAndCleanBody(response.Body)
+	if err != nil {
+		return netip.Addr{}, fmt.Errorf("reading response: %w", err)
+	}
+
+    if response.StatusCode != http.StatusOK {
+        return netip.Addr{}, fmt.Errorf("%w: %d: %s",
+			errors.ErrHTTPStatusNotValid, response.StatusCode, utils.ToSingleLine(s))
+    }
```

Comment:
is there any useful information in the body? if it's just for the error case, let's use instead:

```suggestion
    if response.StatusCode != http.StatusOK {
        return netip.Addr{}, fmt.Errorf("%w: %d: %s",
			errors.ErrHTTPStatusNotValid, response.StatusCode, utils.BodyToSingleLine(response.Body))
    }
```

Reply:
the body looks something like
```json
{
    "records": [
        {
            "id": "aa051759-72e6-4034-8af5-dafba3eebb79",
            "data": "127.1.2.3",
            "name": "",
            "priority": 0,
            "ttl": 300,
            "type": "A",
            "comment": null
        }
    ]
}
```

I think there is nothing useful, let me know if you disagree. Otherwise I will apply your change.

Reply:
What happens if you send a request with the wrong secret key or an owner that doesn't exist, what's the json body you get in the response? It might be a standardized error format which we could parse and use to build a better error to log to the user 😉  Not compulsory, but nice to have if you have the time

Reply:
if the secret key is wrong the response is:
```json
{
    "message": "authentication is denied",
    "method": "api_key",
    "reason": "invalid_argument",
    "type": "denied_authentication"
}
```

If the owner doesn't exist then the record with the right domain is added. The response is indistinguishable from one where the owner already exists.

So, I changed the code into:
```go
if response.StatusCode != http.StatusOK {
		var errorResponse struct {
			Message string `json:"message"`
			Type    string `json:"type"`
		}
		if jsonErr := json.Unmarshal([]byte(s), &errorResponse); jsonErr == nil {
			if errorResponse.Type == "denied_authentication" {
				return netip.Addr{}, fmt.Errorf("%w: %s", errors.ErrAuth, errorResponse.Message)
			}
		}
		return netip.Addr{}, fmt.Errorf("%w: %d: %s",
			errors.ErrHTTPStatusNotValid, response.StatusCode, utils.ToSingleLine(s))
	}
```

Is that ok?